File: common/src/main/scala/edu/cdl/iot/common/schema/Schema.scala
package edu.cdl.iot.common.schema

import java.util.UUID

import edu.cdl.iot.common.schema.Field.{getClassifications, withFeatureType}
import edu.cdl.iot.common.schema.enums.{FieldClassification, FieldType, PartitionScheme}
import edu.cdl.iot.common.schema.enums.FieldType.FeatureType
import edu.cdl.iot.common.yaml.SchemaYaml
import org.joda.time.DateTime
import org.yaml.snakeyaml.Yaml

import collection.JavaConverters._
import scala.collection.mutable

case class Schema(yaml: SchemaYaml) {
  val org: String = yaml.org
  val orgGuid: UUID = UUID.fromString(yaml.orgGuid)
  val name: String = yaml.name
  val projectGuid: UUID = UUID.fromString(yaml.projectGuid)
  val fields: List[Field] = yaml.fields.asScala.toList.map(x => Field(x))
  val dataSources: List[DataSource] = yaml.dataSources.asScala.toList.map(DataSource)
  val importOptions: ImportOptions = ImportOptions(yaml.importOptions)
  val partitionScheme: PartitionScheme.Value = PartitionScheme.fromString(yaml.partitionScheme)
  val featureType: FeatureType = FieldType.fromString(yaml.featureType)

  def toYaml: String = (new Yaml).dumpAsMap(yaml)

  def getPartitionString(date: DateTime): String = {
    val pattern = partitionScheme match {
      case PartitionScheme.DAY => "yyyy-MM-dd"
      case PartitionScheme.HOUR => "yyyy-MM-dd HH"
      case PartitionScheme.MINUTE => "yyyy-MM-dd HH:mm"
    }
    date.toString(pattern)
  }

  def getNextPartition(date: DateTime): DateTime = {
    partitionScheme match {
      case PartitionScheme.DAY => date.plusDays(1)
      case PartitionScheme.HOUR => date.plusHours(1)
      case PartitionScheme.MINUTE => date.plusMinutes(1)
    }
  }

  def getPartitionsInRange(from: DateTime, to: DateTime): List[String] = {
    val partitions = mutable.Set[String]()
    var cursor = from
    while (to.getMillis >= cursor.getMillis) {
      partitions.add(getPartitionString(cursor))
      cursor = getNextPartition(cursor)
    }

    partitions.toList
  }

  def getFileName: String = (if (importOptions == null || importOptions.fileName == null) name else s"$name.csv").toLowerCase()

  def getKey(row: Array[String]): String =
    getClassifications(fields.zipWithIndex, FieldClassification.Key)
      .map(tuple => row(tuple._2))
      .mkString("_")

  def getTimestamp(row: Array[String]): String =
    getClassifications(fields.zipWithIndex, FieldClassification.Timestamp)
      .map(tuple => row(tuple._2))
      .mkString("")


  def getFeatures(row: Array[String]): Map[String, String] =
    getClassifications(fields.zipWithIndex, FieldClassification.Feature)
      .filter(tuple => tuple._2 < row.length)
      .map(tuple => (tuple._1.name.toLowerCase, row(tuple._2)))
      .toMap

  def getFeaturesWithType(row: Array[String], featureType: FeatureType): Map[String, String] =
    withFeatureType(
      getClassifications(fields.zipWithIndex, FieldClassification.Feature),
      featureType)
      .filter(tuple => tuple._2 < row.length)
      .map(tuple => (tuple._1.name.toLowerCase, row(tuple._2)))
      .toMap

  def getLabels(row: Array[String]): Map[String, String] =
    getClassifications(fields.zipWithIndex, FieldClassification.Label)
      .filter(tuple => tuple._2 < row.length)
      .map(tuple => (tuple._1.name.toLowerCase, row(tuple._2)))
      .toMap

  // TODO We will want to create actual type checks for this
  def validate(row: Array[String]): Boolean = fields.length == row.length

  def getFields: Array[Field] = fields.toArray

  def getImportOptions: ImportOptions = importOptions

  def getPartitionScheme: String = partitionScheme.toString

  def getFeatureType: String = featureType.toString
}

File: training/src/main/python/util/Schema.py
import string
from datetime import datetime, timedelta

import yaml

from dao.TrainingDao import TrainingDao


class ImportOptions:
    def __init__(self, yaml: dict):
        self.includes_header = yaml['includesHeader']


class Field:
    def __init__(self, yaml: dict):
        self.name = yaml['name']
        self.type = yaml['type']
        self.classification = yaml['classification']


class Schema:
    def __init__(self, schema_yaml: string):
        dict = yaml.load(schema_yaml)
        self.name = dict['name']
        self.project_guid = dict['projectGuid']
        self.partition_scheme = dict['partitionScheme']
        self.import_options = ImportOptions(dict['importOptions'])
        self.feature_type = dict['featureType']
        self.fields = list(map(lambda x: Field(x), dict['fields']))
        self.org_guid = dict['orgGuid']

    def __repr__(self) -> string:
        return "name:%s guid:%s" % (self.name, self.project_guid)

    def prediction_label(self) -> string:
        results = list(filter(lambda x: x.classification == "Label", self.fields))
        return results[0].name.lower()

    def get_partition_key(self, date: datetime) -> string:
        if self.partition_scheme == "DAY":
            return date.strftime("%Y-%m-%d")
        elif self.partition_scheme == "HOUR":
            return date.strftime("%Y-%m-%d %H")
        elif self.partition_scheme == "MINUTE":
            return date.strftime("%Y-%m-%d %H:%M")
        else:
            return date.strftime("%Y-%m-%d")

    def get_next_partition(self, date: datetime) -> datetime:
        if self.partition_scheme == "DAY":
            return date + timedelta(days=1)
        elif self.partition_scheme == "HOUR":
            return date + timedelta(hours=1)
        elif self.partition_scheme == "MINUTE":
            return date + timedelta(minutes=1)
        else:
            return date + timedelta(days=1)

    def get_partitions_in_range(self, start: datetime, end: datetime) -> list:
        cursor = start
        partitions = []
        while cursor <= end:
            partitions.append(self.get_partition_key(cursor))
            cursor = self.get_next_partition(cursor)
        return partitions


class SchemaFactory:
    def __init__(self, dao: TrainingDao):
        self.dao = dao

    def getSchema(self, project_guid: string) -> Schema:
        (schema_yaml, org_guid) = self.dao.get_schema(project_guid)
        return Schema(schema_yaml)
